### PR TITLE
Raise warning if torch.compile options change without reset

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94680

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: test/inductor/test_config.py [hunk 1]
```diff
@@ -118,11 +118,65 @@ def test_compile_api(self):
...
+        self.assertRaises(torch._dynamo.exc.ResetRequired, lambda: a(torch.randn(10)))
+
+        with torch._dynamo.config.patch(
+            raise_on_backend_change=False
+        ), self.assertWarns(Warning):
+            # normally it is just a warning
+            a(torch.randn(10))
+
+        # only warn once
+        a(torch.randn(10))
```
Comment (wconstab) [inline test/inductor/test_config.py:178]: do you want to assert not warns here?

File: torch/__init__.py [hunk 2]
```diff
@@ -1320,11 +1320,8 @@ class _TorchCompileInductorWrapper:
     compiler_name = "inductor"
 
     def __init__(self, mode, options, dynamic):
-        from torch._inductor.compile_fx import compile_fx
-
-        self.compile_fn = compile_fx
-        self._torchdynamo_orig_callable = compile_fx
         self.config = dict()
+        self.dynamic = dynamic
```
Comment (wconstab) [inline torch/__init__.py:1324]: why is dynamic not in config?

**Review by wconstab — APPROVED** [review]

File: test/inductor/test_config.py (hunk 1 above)
Comment (jansel) [reply test/inductor/test_config.py:178]: It won't warn because it already warned on L175

File: torch/__init__.py (hunk 2 above)
Comment (jansel) [reply torch/__init__.py:1324]: Config here is inductor config, and dynamic is not part of inductor's config.  (It controls dynamo behavior.)

**Review by malfet — APPROVED** [review]

**Pull request closed without merging**